vehicle/src/Vehicle/Backend/LossFunction/Logics.hs

```diff
@@ -0,0 +1,210 @@
+
+-- definiton of the LExpr - all expressions allowed in a loss constraint
+
+data LExpr
+  = Negation LExpr                           -- this is minus, not the logical operation of negation
+  | Constant Double                           -- constant
+  | Min LExpr LExpr                           -- Min
+  | Max LExpr LExpr                           -- Max
+  | Addition LExpr LExpr                      -- addition
+  | Subtraction LExpr LExpr                   -- subtraction
+  | Multiplication LExpr LExpr                -- multiplication
+  | Division LExpr LExpr                      -- division
+  | IndicatorFunction LExpr LExpr             -- an indicator function
+  | Variable V.DBIndex                        -- variable (bound)
+  | FreeVariable Name                         -- variable (free)
+  | NetworkApplication Name (NonEmpty LExpr)  -- neural network
+  | Quantifier Quantifier Name Domain LExpr   -- quantifiers forall, exists
+  | At LExpr LExpr                            -- at
+  | TensorLiteral [LExpr]                     -- tensor
+  | Lambda Name LExpr                         -- lambda expression
+  | Let Name LExpr LExpr                      -- let expression
+  | Power LExpr LExpr                         -- exponential
+  deriving (Eq, Ord, Generic, Show)
+
+--------------------------------------------------------------------------------
+-- other Definitions
+
+data Quantifier
+  = All
+  | Any
+  deriving (Eq, Ord, Generic, Show)
+
+newtype Domain = Domain ()
+  deriving (Eq, Ord, Generic, Show)
+
+-- template for different avilable differentiable logics
+
+data DifferentialLogicImplementation = DifferentialLogicImplementation
+  { compileAnd          :: LExpr -> LExpr -> LExpr
+  , compileOr           :: LExpr -> LExpr -> LExpr
+  , compileNot          :: Maybe (LExpr -> LExpr)
+  , compileImplies      :: LExpr -> LExpr -> LExpr
+
+  , compileLe           :: LExpr -> LExpr -> LExpr
+  , compileLt           :: LExpr -> LExpr -> LExpr
+  , compileGe           :: LExpr -> LExpr -> LExpr
+  , compileGt           :: LExpr -> LExpr -> LExpr
+  , compileEq           :: LExpr -> LExpr -> LExpr
+  , compileNeq          :: LExpr -> LExpr -> LExpr
+
+  , compileTrue         :: Double
+  , compileFalse        :: Double
+    }
+
+
+--------------------------------------------------------------------------------
+-- different avilable differentiable logics (types of translation from the constraint to loss function) are
+-- DL2                              --Fischer, Marc, et al. "Dl2: Training and querying neural networks with logic."  PMLR, 2019.
+-- Godel                            --van Krieken, et al. "Analyzing differentiable fuzzy logic operators." Artificial Intelligence 302 (2022): 103602.
+-- Lukasiewicz                      --as above
+-- Product based                    --as above
+-- Yager                            --as above
+
+--they can be found in Vehicle.Backend.Prelude and the default option if none is provided is DL2.
```

Comment:
I'm not sure documenting this here is that useful. Is this documented on the command line interface? That would be a better place for it?